### DISCUSSION TITLE
Bitcast array from int32 to int64

### DISCUSSION THREAD

**Discussion by juliandwain** [jax-ml/jax#8784] [Q&A] [answered]
Hi,
I'm trying to replicate the behavior of [tensorflow's bitcast function](https://www.tensorflow.org/api_docs/python/tf/bitcast).
I have an array of type ``int32`` and I want to convert it to ``int64``.
I found that jax has the [``jax.lax.bitcast_convert_type``](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.bitcast_convert_type.html) function, but this does not work properly.
When I am wrapping my operation inside [``jax.experimental.enable_x64``](https://jax.readthedocs.io/en/latest/jax.experimental.html), I get the following error:
```
Exception has occurred: TypeError
`bitcast_convert_type` for operand type (int32) must have destination type (int64) of same size.
  File "/home/user/proj/gendata.py", line 47, in function
    e= jax.lax.bitcast_convert_type(
  File "/home/user/proj/gendata.py", line 60, in <module>
    a, b = function1(inputs)
```
For context:
- I have an input array ``a`` with shape ``(100,2)`` with dtype ``int32``
- according to the [tensorflow documentation](https://www.tensorflow.org/api_docs/python/tf/bitcast), ``tf.bitcast``, "If ``T`` is smaller than ``type``, the operator requires that the rightmost dimension be equal to sizeof(``type``)/sizeof(``T``). The shape then goes from [..., sizeof(``type``)/sizeof(``T``)] to [...].", in my case, the array is flattened
- the wrapping is done like such:
```python
with jax.experimental.enable_x64():
  a = jax.lax.bitcast_convert_type(...)
```

As I am currently pretty much stuck here, I appreciate every little help I get!
Essentially, I am trying to replicate ``tf.bitcast`` behavior in jax.

Thanks in advance!

Best,
J

**Comment by jakevdp** [comment, marked as answer]
One way to do this is with the `view` method of JAX arrays. For example:
```python
import jax
jax.config.update('jax_enable_x64', True)

import jax.numpy as jnp
x32 = jnp.arange(20, dtype='int32').reshape(10, 2)
x64 = x32.view('int64').ravel()
print(x64)
# [ 4294967296 12884901890 21474836484 30064771078 38654705672 47244640266
#  55834574860 64424509454 73014444048 81604378642]

import tensorflow as tf
x32 = tf.reshape(tf.range(20, dtype='int32'), (10, 2))
x64 = tf.bitcast(x32, 'int64')
print(x64)
# tf.Tensor(
# [ 4294967296 12884901890 21474836484 30064771078 38654705672 47244640266
#  55834574860 64424509454 73014444048 81604378642], shape=(10,), dtype=int64)
```

**Reply by juliandwain** [reply]
Thank you so much for the really quick answer!
It worked perfectly fine!

I am now trying to use ``tf.bitcast`` the other way around, i.e., changing my dtype from ``int64`` back to ``int32`` (after doing some operations inbetween). Using ``tf.bitcast`` would give me an array with shape ``(*, 2)`` back again.

I am correct that I obtain this in JAX by:

```python
x32_new = x64.view("int32").reshape((-1, 1)) 
```

**Reply by jakevdp** [reply]
That will give you a `(*, 1)` array. You could get the original back with `x64[:,, None].view('int32')`

**Reply by juliandwain** [reply]
```python
import jax

jax.config.update("jax_enable_x64", True)

import numpy as np
import jax.numpy as jnp

n = 20
seed = 12345

rng = np.random.default_rng(seed=seed)
arr = np.arange(n, dtype="int32").reshape(n // 2, 2)
arr = rng.permutation(arr)

x32j = jnp.asarray(arr)
x64j = x32j.view("int64").ravel()
print(x64j)
# [38654705672 73014444048 12884901890 30064771078 64424509454 81604378642
# 55834574860  4294967296 21474836484 47244640266]
uniquej = jnp.unique(x64j)
x32j = uniquej.view("int32").reshape((-1, 2))
print(x32j) # not the same as tf version
# [[ 0  1]
 # [ 2  3]
 # [ 4  5]
 # [ 6  7]
 # [ 8  9]
 # [10 11]
 # [12 13]
 # [14 15]
 # [16 17]
 # [18 19]]

import tensorflow.compat.v1 as tf

x32tf = tf.convert_to_tensor(arr)
x64tf = tf.bitcast(x32tf, "int64")
print(x64tf)
# [38654705672 73014444048 12884901890 30064771078 64424509454 81604378642
# 55834574860  4294967296 21474836484 47244640266], shape=(10,), dtype=int64)
uniquetf = tf.unique(x64tf)[0]  # different return behavior than numpy
x32tf = tf.bitcast(uniquetf, tf.int32)
print(x32tf)
# [[ 8  9]
# [16 17]
# [ 2  3]
# [ 6  7]
# [14 15]
# [18 19]
# [12 13]
# [ 0  1]
# [ 4  5]
# [10 11]]

```

**Reply by juliandwain** [reply]
> That will give you a `(*, 1)` array. You could get the original back with `x64[:,, None].view('int32')`

Yes sorry that was a typo from my side, I meant 2 not 1

**Reply by juliandwain** [reply]
I think the issue is within the ``unique`` call, since the return values are different for JAX and TF:

JAX
```python
[ 4294967296 12884901890 21474836484 30064771078 38654705672 47244640266
 55834574860 64424509454 73014444048 81604378642]
```

TF
```python
[38654705672 73014444048 12884901890 30064771078 64424509454 81604378642
 55834574860  4294967296 21474836484 47244640266]
```

They contain the same values but the ordering is different.

When executing

```python
_arr = jnp.asarray(
    [
        38654705672,
        73014444048,
        12884901890,
        30064771078,
        64424509454,
        81604378642,
        55834574860,
        4294967296,
        21474836484,
        47244640266,
    ]
)
print(_arr[:, None].view("int32"))
# [[ 8  9]
 # [16 17]
 # [ 2  3]
 # [ 6  7]
 # [14 15]
 # [18 19]
 # [12 13]
 # [ 0  1]
 # [ 4  5]
 #[10 11]]
```

the output is exactly the same as the one from tensorflow.

**Reply by jakevdp** [reply]
RIght, `jnp.unique` is designed to have the same output as `np.unique`, which sorts its outputs.

**Reply by juliandwain** [reply]
Is it possible to make it similar to TF?

**Reply by jakevdp** [reply]
I think this should give the same result as `tf.unique`, though I haven't tested it extensively:
```python
x = jnp.array([4, 2, 4, 3, 3, 2, 4])
x_unique, i = jnp.unique(x, return_index=True)
print(x_unique[jnp.argsort(i)])
# [4 2 3]
```
